src/matrix_projections.jl

```diff
@@ -396,12 +398,41 @@ function image_basis!(A::AbstractSparseMatrix)
     return A, p
 end
 
-function image_basis!(A::AbstractMatrix{T}) where {T<:Union{AbstractFloat,Complex}}
+function _orth!(M::AbstractMatrix{T}) where T<:Union{AbstractFloat, Complex}
+    F = svd!(convert(Matrix{T}, M))
+    M_rank = let m = T <: Complex ? 2eps(real(T)) : eps(T)
+        sum(F.S .> maximum(size(M)) * m)
+    end
+    return F.Vt, M_rank
+end
+
+function _orth!(M::AbstractSparseMatrix{T}) where T<:Union{AbstractFloat, Complex}
+    F = qr(M)
+    M_rank= rank(F)
+    result = let tmp = F.Q * Matrix(I, size(F.Q, 2), M_rank)
```

Comment:
If we change to `size(F.Q, 1)` then `_orth!(M)` technically returns (numerical) orthonormal basis for column space of `M` (independent of whether it is square). Here it is fine since we will call it for square `M` corresponding to sparse orthogonal projection onto isotypic components. Also, should this function be `_orth` (without bang)?

Reply:
As far as I understand You can't multiply `n×m` matrix by a `n×r` one?
I added `!` for uniformity (the one for dense matrices modifies its arguments), `_orth` would be just fine, it's an internal function ;)

Anyway bang just means that the function may modify its arguments (usually the first one), but in principle one should not rely on mutation, but return the modified result

Reply:
As discussed, you are correct: to get first `M_rank` columns of `F.Q` we need to multiply with identity of size `(size(F.Q, 2), M_rank)`.